src/components/task-step/exercise.cjsx

```diff
@@ -1,67 +1,205 @@
 
-  canOnlyContinue: ->
-    {id} = @props
-    _.chain(StepPanel.getRemainingActions(id))
-      .difference(['clickContinue'])
-      .isEmpty()
-      .value()
+  getInitialState: ->
+    partsInfo = @getPartsInfo()
+    taskInfo = @getTaskInfo()
+
+    _.extend({}, partsInfo, taskInfo)
+
+  getPartsInfo: (props) ->
+    props ?= @props
+
+    {id, taskId, courseId, onNextStep} = props
+    parts = TaskStore.getStepParts(taskId, id)
+
+    lastPartId = _.last(parts).id
+    isSinglePartExercise = @isSinglePart(parts)
+
+    {parts, lastPartId, isSinglePartExercise}
+
+  getTaskInfo: (props) ->
+    props ?= @props
+    {taskId} = props
 
-  render: ->
-    {id, taskId} = @props
-    step = TaskStepStore.get(id)
     task = TaskStore.get(taskId)
-    stepIndex = TaskStore.getStepIndex(taskId, id)
-
-    waitingText = switch
-      when TaskStepStore.isLoading(id) then "Loading…"
-      when TaskStepStore.isSaving(id)  then "Saving…"
-      else null
-
-    getReadingForStep = (id, taskId) ->
-      TaskStore.getReadingForTaskId(taskId, id)
-
-    getCurrentPanel = (id) ->
-      unless TaskStepStore.isSaving(id)
-        currentPanel = StepPanel.getPanel(id)
-
-    controlText = 'Continue' if task.type is 'reading' and @canOnlyContinue()
-
-    <div className='exercise-wrapper' data-step-number={stepIndex + 1}>
-      <Exercise
-        {...@props}
-        freeResponseValue={step.temp_free_response}
-        controlText={controlText}
-        step={step}
-        footer={<StepFooter/>}
-        waitingText={waitingText}
-
-        canTryAnother={TaskStepStore.canTryAnother(id, task)}
-        isRecovering={TaskStepStore.isRecovering(id)}
-        disabled={TaskStepStore.isSaving(id)}
-        canReview={StepPanel.canReview(id)}
-        isContinueEnabled={StepPanel.canContinue(id)}
-
-        getCurrentPanel={getCurrentPanel}
-        getReadingForStep={getReadingForStep}
-        setFreeResponseAnswer={TaskStepActions.setFreeResponseAnswer}
-        onFreeResponseChange={@updateFreeResponse}
-        freeResponseValue={TaskStepStore.getTempFreeResponse(id)}
-        setAnswerId={TaskStepActions.setAnswerId}
-      />
-    </div>
+    currentStep = TaskProgressStore.get(taskId)
+
+    {task, currentStep}
+
+  isSinglePart: (parts) ->
+    parts.length is 1
+
+  componentWillMount: ->
+    @startListeningForProgress()
+
+  componentWillUnmount: ->
+    @stopListeningForProgress()
+
+  componentWillReceiveProps: (nextProps) ->
+    if nextProps.taskId isnt @props.taskId
+      nextState = _.extend({}, @getTaskInfo(nextProps), @getPartsInfo(nextProps))
+      @setState(nextState)
+      @stopListeningForProgress()
+      @startListeningForProgress(nextProps)
+    else
+      @setState(@getPartsInfo(nextProps))
+
+  stopListeningForProgress: (props) ->
+    props ?= @props
+    {taskId} = props
+
+    TaskProgressStore.off("update.#{taskId}", @setCurrentStepFromProgress)
+
+  startListeningForProgress: (props) ->
+    props ?= @props
```

Comment:
when `@ stopListeningForProgress` is called without an argument, the`props` is `null` and then defaults to `@props`.  this makes it easy to call the same function with `nextProps` or just leave it empty so that it uses `@props`